lib/broadway/producer.ex

```diff
@@ -105,21 +123,18 @@ defmodule Broadway.Producer do
   end
 
   @impl true
-  def handle_demand(demand, state) do
-    %{module: module, transformer: transformer, module_state: module_state} = state
-
-    case module.handle_demand(demand, module_state) do
-      {:noreply, events, new_module_state} when is_list(events) ->
-        messages = transform_events(events, transformer)
-        {:noreply, messages, %{state | module_state: new_module_state}}
+  def handle_demand(demand, state)
 
-      {:noreply, events, new_module_state, :hibernate} ->
-        messages = transform_events(events, transformer)
-        {:noreply, messages, %{state | module_state: new_module_state}, :hibernate}
+  # If we're rate limited, we store the demand in the buffer instead of forwarding it.
+  # We'll forward it once the rate limit is lifted.
+  def handle_demand(demand, %{rate_limiting: %{state: :closed}} = state) do
+    state = update_in(state.rate_limiting.demand_buffer, &:queue.in(demand, &1))
+    {:noreply, [], state}
+  end
 
-      {:stop, reason, new_module_state} ->
-        {:stop, reason, %{state | module_state: new_module_state}}
-    end
+  def handle_demand(demand, state) do
```

Comment:
```suggestion
  @impl true
  def handle_demand(demand, state) do
```

Reply:
There's a bodiless clause above

```elixir
@impl true
def handle_demand(demand, state)
```

so that we can use `@impl true` only once. I usually do this but can change to multiple uses of `@impl true` if you prefer.

Reply:
I believe I started using `@impl true` on all clauses as it helps catch a wrong definition in a clause too. So we can do it here and remove the bodiless one! :+1: